### PR TITLE
feat: conversion of openapi '3.0' to asyncapi '3.0'

### LINKED ISSUE #233: Convert OpenAPI 3.0 to AsyncAPI 3.0
This issue is part of a longer vision for the converter that can be found here: https://github.com/asyncapi/converter-js/issues/221

It would be great to open up how easily people can switch or try out AsyncAPI based on existing resources. One of these is OpenAPI 3.0: https://spec.openapis.org/oas/v3.0.3. With the new AsyncAPI v3, we should be able to map this 100% from OpenAPI to AsyncAPI.

**In order to solve this it requires a lot of conversion that might not be straightforward. While this issue is tagged as a `good first issue` it requires a lot of effort to push through, on the bright side, solving this issue will help you gain knowledge about OpenAPI and AsyncAPI, and how they relate and differentiate. Of course alongside a huge deal of TS, testing, documentation, and working in Open Source in general.**

To make it easier to digest, this issue can be split up into multiple sections to simplify the process of solving this:

1. Step: Convert the following OpenAPI keywords to form a basic AsyncAPI document with no channels nor operations: `openapi`, `info`, `servers`, `tags`, `externalDocs`, and `security`.
2. Step: Convert the OpenAPI [`paths` keyword](https://spec.openapis.org/oas/v3.0.3#paths-object) into channels and operations in AsyncAPI
3. Step: Convert the [OpenAPI components](https://spec.openapis.org/oas/v3.0.3#components-object) into AsyncAPI components

**Potential next steps that should NOT be focused on right now**

4. Step: Enable multiple options to customize the conversion.
5. Step: Enable the split of the OpenAPI document into one or more AsyncAPI documents.
6. Step: Integrate it into the AsyncAPI CLI to enable faster conversions for users.
7. Write a blog post about this new conversion

> NOTICE, because of the difficulty of this issue, you will be able to contact me all the time through Slack if you get stuck solving it at any point and even have calls when relevant.

### PR DESCRIPTION
**Pull request by Gmin2**
**Description**

- [x] Converted `openapi`, `info`, `servers`, `tags`, `externalDocs,` and `security`.
- [x] Converted `path` keyword
- [x] Converted `components`.

**Related issue(s)**
Fixes #233

### CODE REVIEW

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Your code is way too tangled together at the moment. There should be clear separations between converting AsyncAPI between versions and converting OpenAPI to AsyncAPI 🙂 

Its completely fine to have 2 functions, one called `convert` for converting AsyncAPI between versions and then another called `?` to convert OpenAPI to AsyncAPI 🤙  

Also make sure you add more test, especially edge cases for it to convert 🙂 

But this is a good first PR to merge, good first initial feature 👍

**Comment by Gmin2** [comment]
dont merge @jonaslagoni need to add some more test cases

**Comment by aeworxet** [comment]
@asyncapi/bounty_team

File: src/openapi.ts
Comment (jonaslagoni) [inline src/openapi.ts:24]: Since we are converting from one document type (OpenAPI) to another,  would it not make sense to create an empty AsyncAPI object and apply it to that instead of changing the openapi object?

File: test/output/openapi-to-asyncapi/no-channel-parameter.yml
Comment (jonaslagoni) [inline test/output/openapi-to-asyncapi/no-channel-parameter.yml:38]: This is not a property in AsyncAPI.

File: test/output/openapi-to-asyncapi/no-channel-parameter.yml
Comment (jonaslagoni) [inline test/output/openapi-to-asyncapi/no-channel-parameter.yml:22]: URL is split up in AsyncAPI v3: https://www.asyncapi.com/docs/reference/specification/v3.0.0#serverObject

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Make sure to double-check whether the converted AsyncAPI document is valid 🙂

File: test/output/openapi-to-asyncapi/no-channel-parameter.yml
Comment (ibishal) [reply test/output/openapi-to-asyncapi/no-channel-parameter.yml:22]: it was done only, i have written the testcase wrongly

Comment (Gmin2) [reply test/output/openapi-to-asyncapi/no-channel-parameter.yml:22]: it was done only, i had forgotten to update the test case

**Comment by jonaslagoni** [comment]
@Gmin2 ping me if you need anything

**Comment by Gmin2** [comment]
@jonaslagoni can we merge this PR, i will make another  PR with more testcases and refining the conversion of path objects

**Comment by jonaslagoni** [comment]
Since you already included everything in one PR and it is not a fully working change by it self, we cannot unfortunately...

**Comment by Gmin2** [comment]
@jonaslagoni completed all the work, please review

**Comment by jonaslagoni** [comment]
@Gmin2 let me know when you are ready for a final review.

**Comment by jonaslagoni** [comment]
@Gmin2 haha, okay 👍

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Looking good, a bit more is needed before we can merge this :v: 

A few things are needed:
- Make sure you test the perspective option
- Make sure that new functions have jsdocs associated that explains abstracted what it does
- Make sure that the converted operation object also assigns the `.reply` section where needed.

**Comment by jonaslagoni** [comment]
@Gmin2 also make sure you update the readme file to introduce the new feature

File: src/openapi.ts
Comment (Gmin2) [reply src/openapi.ts]: updated at [recent commit](https://github.com/asyncapi/converter-js/pull/269/commits/7a5ee2a6e030f3384602a9a29f03297e09539153)

File: src/openapi.ts
Comment (ibishal) [reply src/openapi.ts]: i have set the schemaFormat to `application/vnd.oai.openapi;version=3.0.0`

Comment (Gmin2) [reply src/openapi.ts]: i have set the schemaFormat to `application/vnd.oai.openapi;version=3.0.0`

File: src/convert.ts
Comment (jonaslagoni) [inline src/convert.ts:71]: This is gonna be a user error, lets use something they understand not internal descriptions. Maybe something like:
`throw new Error(\`We are not able to convert OpenAPI ${converterVersion} to AsyncAPI, please raise a feature request.\`);`

File: test/output/openapi-to-asyncapi/operation_and_parameter_client.yml
Comment (jonaslagoni) [inline test/output/openapi-to-asyncapi/operation_and_parameter_client.yml:79]: Not sure what to do here. Technically by not defining `messages`, all messages within `#/channels/items` is assumed to be possible...

I think we might have to define an empty list here 🤔

File: src/openapi.ts
Comment (jonaslagoni) [inline src/openapi.ts:494]: Is this needed anymore?

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Forgot to mention (actually thought I did in the last review) remember to change the main readme file in the repo so everyone can find this new functionality 🙂

File: src/openapi.ts
Comment (jonaslagoni) [reply src/openapi.ts]: Yep 👍

File: test/output/openapi-to-asyncapi/operation_and_parameter_client.yml
Comment (ibishal) [reply test/output/openapi-to-asyncapi/operation_and_parameter_client.yml:79]: I have define an empty list when there are no specific `messages` in `operation`

Comment (Gmin2) [reply test/output/openapi-to-asyncapi/operation_and_parameter_client.yml:79]: I have define an empty list when there are no specific `messages` in operation

File: src/openapi.ts
Comment (Gmin2) [reply src/openapi.ts]: Yes we dont need this anymore, updated it

**Comment by Gmin2** [comment]
> Forgot to mention (actually thought I did in the last review) remember to change the main readme file in the repo so everyone can find this new functionality 🙂

Done, need a final look @jonaslagoni

File: src/convert.ts
Comment (Gmin2) [reply src/convert.ts]: Done

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
I know this last efforts always are the toughest to perfect but makes all the difference for the users 🤙 

I see some edge cases that are not covered so far by the tests and will make complex OpenAPI documents fail. I dont think we can do anything about `External -> internal`, except make a notice in the readme about it being something not supported. However, `External schema objects` and `Nested schema objects` is something that should be supported. Same with parameters of course.

- External schema objects 
```
openapi: 3.0.0
info:
  title: test
  version: 1.0.0
  description: Test
paths:
  /test:
    get:
      responses:
        '200':
          description: Successful response
          content:
            application/json:
              schema:
                $ref: './external.json'
```
- Nested schema objects 
```
openapi: 3.0.0
info:
  title: test
  version: 1.0.0
  description: Test
paths:
  /test:
    get:
      responses:
        '200':
          description: Successful response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/test'
components:
  schemas:
    test:
      type: object
      properties:
        message:
          $ref: '#/components/schemas/test2'
    test2:
      type: object
      properties:
        message:
          type: string
```
Here, you cannot convert both schemas to the multi format schema, because referencing test2 as a multi format schema inside test is not allowed.

- External -> internal
```yml
openapi: 3.0.0
info:
  title: test
  version: 1.0.0
  description: Test
paths:
  /test:
    get:
      responses:
        '200':
          description: Successful response
          content:
            application/json:
              schema:
                $ref: './external.json'
components:
  schemas:
    test:
      type: object
      properties:
        message:
          type: string

# external.json
type: object
properties:
  message:
    $ref: './asyncapi.json#/components/schemas/test'
```

Same as above, referencing test as a multi format schema inside external is not allowed.
- Parameters are not being converted in the AsyncAPI document, even though we have explicit test document for it 🤔

File: test/output/openapi-to-asyncapi/callbacks_and_contents.yml
Comment (ibishal) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: Hey @jonaslagoni, if the `schema.type` will be `object` then it should be openAPI schema only 
image
 when i try to convert it to multiFormat 
 ```
 payload:
          schemaFormat: application/vnd.oai.openapi;version=3.0.0
          schema:
            type: object
            properties:
              file:
                type: string
                format: binary
 ```
 
 it gives error in studio

File: test/output/openapi-to-asyncapi/callbacks_and_contents.yml
Comment (Gmin2) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:31]: Hey @jonaslagoni, if the `schema.type` will be `object` then it should be openAPI schema only 
image
 when i try to convert it to multiFormat 
 ```
 payload:
          schemaFormat: application/vnd.oai.openapi;version=3.0.0
          schema:
            type: object
            properties:
              file:
                type: string
                format: binary
 ```
 
 it gives error in studio

File: test/output/openapi-to-asyncapi/callbacks_and_contents.yml
Comment (Gmin2) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: Hey @jonaslagoni, if the `schema.type` will be `object` then it should be openAPI schema only 
image
 when i try to convert it to multiFormat 
 ```
 payload:
          schemaFormat: application/vnd.oai.openapi;version=3.0.0
          schema:
            type: object
            properties:
              file:
                type: string
                format: binary
 ```
 
 it gives error in studio

Comment (jonaslagoni) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: Must be a bug in Studio then 🤔

Comment (jonaslagoni) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: What error does it give in Studio?

Comment (Gmin2) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: image

it gives this error
```
Core schema meta-schema

Source: asyncapi.json
"payload" property must match "then" schema
```

Comment (jonaslagoni) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: Try make a bug report in Studio, i will be afk for the next two weeks, so maybe try on slack as well

Comment (Gmin2) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: should i push my other changes that you have mentioned, and leave the `payload` schema as it is

Comment (jonaslagoni) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: Ignore Studio, just double check the spec that its allowed.

That change is needed to complete the implementation.

Comment (ibishal) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: The [specs](https://asyncapi.com/definitions/3.0.0/multiFormatSchema.json) says apart from type `object` it should be fine

Comment (Gmin2) [reply test/output/openapi-to-asyncapi/callbacks_and_contents.yml:24]: The [specs](https://asyncapi.com/definitions/3.0.0/multiFormatSchema.json) says apart from type `object` it should be fine

File: src/openapi.ts
Comment (jonaslagoni) [inline src/openapi.ts:473]: This is not entirely correct we still need to support External schema objects case, as we still need to convert the schema to multi schema object even if external. We just dont touch the external file and do any changes there. That is the restriction.


Make sure to add a test case for this.

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Small comments :v:

File: src/openapi.ts
Comment (Gmin2) [reply src/openapi.ts]:
```ts
if (schema.$ref.startsWith('./') || schema.$ref.startsWith('http')) {
      // Convert external references to multi-format schema objects
      return {
        schemaFormat: 'application/vnd.oai.openapi;version=3.0.0',
        schema: schema
      };
    }
    return schema;
```
pushed this

**Review by jonaslagoni — COMMENTED** [review]
Nice, good first step 🤙 

Lets give the other maintainers a chance to do final review before merging 😊

**Comment by jonaslagoni** [comment]
cc @fmvilas @magicmatatjahu

**Review by jonaslagoni — APPROVED** [review]
Merging as two weeks have passed :v:

Nicely done @Gmin2 💪

**Comment by jonaslagoni** [comment]
@all-contributors please add @Gmin2 for code, doc, example, test

**Pull request merged**
